test/util_test.sh

```diff
@@ -0,0 +1,305 @@
+source "${OS_ROOT}/hack/util.sh"
+source "${OS_ROOT}/test/util.sh"
+os::log::install_errexit
+
+mkdir -p /tmp/openshift/origin/test/cmd/
+JUNIT_OUTPUT_FILE=/tmp/openshift-cmd/junit_output.txt
+
+# set verbosity so we can see that command output renders correctly
+VERBOSE=1
+
+# positive tests
+os::test::expect_success 'exit 0'
+
+os::test::expect_failure 'exit 10'
+
+os::test::expect_success_and_text 'printf "hello" && exit 0' 'hello'
+
+os::test::expect_failure_and_text 'printf "hello" && exit 19' 'hello'
+
+os::test::expect_success_and_not_text 'echo "goodbye" && exit 0' 'hello'
+
+os::test::expect_failure_and_not_text 'echo "goodbye" && exit 19' 'hello'
+
+os::test::expect_code 'exit 195' '195'
+
+os::test::expect_code_and_text 'echo "hello" && exit 213' '213' 'hello'
+
+os::test::expect_code_and_not_text 'echo "goodbye" && exit 213' '213' 'hello'
+
+echo "positive tests: ok"
+
+# negative tests
+
+if os::test::expect_success 'exit 1'; then
+	exit 1
+fi
+
+if os::test::expect_failure 'exit 0'; then
+	exit 1
+fi
+
+if os::test::expect_success_and_text 'echo "goodbye" && exit 0' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_success_and_text 'echo "hello" && exit 1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_success_and_text 'echo "goodbye" && exit 1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_failure_and_text 'echo "goodbye" && exit 1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_failure_and_text 'echo "hello" && exit 0' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_failure_and_text 'echo "goodbye" && exit 0' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_success_and_not_text 'echo "hello" && exit 0' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_success_and_not_text 'echo "goodbye" && exit 1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_success_and_not_text 'echo "hello" && exit 1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_failure_and_not_text 'echo "goodbye" && exit 0' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_failure_and_not_text 'echo "hello" && exit 1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_failure_and_not_text 'echo "hello" && exit 0' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_code 'exit 1' '200'; then
+	exit 1
+fi
+
+if os::test::expect_code_and_text 'echo "hello" && exit 0' '1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_code_and_text 'echo "goodbye" && exit 1' '1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_code_and_text 'echo "goodbye" && exit 0' '1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_code_and_not_text 'echo "goodbye" && exit 0' '1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_code_and_not_text 'echo "hello" && exit 1' '1' 'hello'; then
+	exit 1
+fi
+
+if os::test::expect_code_and_not_text 'echo "hello" && exit 0' '1' 'hello'; then
+	exit 1
+fi
+
+echo "negative tests: ok"
+
+# complex input tests
+
+os::test::expect_success 'echo "hello" | grep hello'
+
+os::test::expect_success 'echo $(echo "hello") | grep hello'
```

Comment:
Add a variable evaluation inside the nesting instead of "hello". This could match even on weird failures.